tests/validators/test_dataclasses.py

```diff
@@ -584,3 +584,180 @@ def validate_b(
     v = SchemaValidator(schema)
     foo = v.validate_python({'a': 1, 'b': b'hello'})
     assert dataclasses.asdict(foo) == {'a': 1, 'b': 'hello world!'}
+
+
+def test_dataclass_self_init():
+    @dataclasses.dataclass(init=False)
+    class Foo:
+        a: str
+        b: bool
+
+        def __init__(self, *args, **kwargs):
+            v.validate_python(ArgsKwargs(args, kwargs), self_instance=self)
+
+    schema = core_schema.dataclass_schema(
+        Foo,
+        core_schema.dataclass_args_schema(
+            'Foo',
+            [
+                core_schema.dataclass_field(name='a', schema=core_schema.str_schema(), kw_only=False),
+                core_schema.dataclass_field(name='b', schema=core_schema.bool_schema(), kw_only=False),
+            ],
+        ),
+    )
+    v = SchemaValidator(schema)
+
+    foo = Foo(b'hello', 'True')
+    assert dataclasses.is_dataclass(foo)
+    assert dataclasses.asdict(foo) == {'a': 'hello', 'b': True}
+
+
+def test_dataclass_self_init_alias():
+    @dataclasses.dataclass(init=False)
+    class Foo:
+        a: str
+        b: bool
+
+    schema = core_schema.dataclass_schema(
+        Foo,
+        core_schema.dataclass_args_schema(
+            'Foo',
+            [
+                core_schema.dataclass_field(name='a', schema=core_schema.str_schema(), validation_alias='aAlias'),
+                core_schema.dataclass_field(name='b', schema=core_schema.bool_schema(), validation_alias='bAlias'),
+            ],
+        ),
+    )
+    v = SchemaValidator(schema)
+
+    def __init__(self, *args, **kwargs):
+        v.validate_python(ArgsKwargs(args, kwargs), self_instance=self)
+
+    Foo.__init__ = __init__
```

Comment:
This test is set up a bit different from above, it would be nice to standardize one way to set up these tests

Reply:
this is almost exactly what we do in pydantic, hence I was trying it.

Reply:
right but I mean we should have some consistency between these tests. E.g. above you access `v` from within `__init__` which works because it's in the function local scope, but here you assign `__init__` onto the already constructed class. No big deal either way, I understood it after reading, but it took me an extra pass.